keps/sig-network/20200921-topology-aware-routing/README.md

```diff
@@ -0,0 +1,413 @@
+![local alg](local-example.png)
+![local alg](local-example2.png)
+
+###### Optimizations
+_Downgrading:_
+1. If failed to maintain an overload less than overload threshold, we downgrade the whole mechanism to the `Balanced` approach.
+![downgrading](opt1.png)
+![downgrading](opt1-cont.png)
+
+_Starting Threshold -- default 3x of zone number:_
+1. As we evaluated algorithms on different datasets, we found that at small scales of endpoints number, it is too risky to conduct a `Prefer` approach. It results in a high probability of EndpointSlice churn (higher ratio change when updating endpoints) and imbalanced traffic load for endpoints. With all that said, we determined a starting threshold based on the test results as 3 times the number of zones. 
+<br>_Note that this starting threshold may vary with different score mechanisms._
+2. Besides the starting threshold, we also want to introduce a padding protection for policy switches. This design is meant to protect unnecessary churn caused by rolling updates. 
+
+Example: starting threshold = 9, padding = 3
+1. Remain `Balanced` approach until # of endpoints hits 12
+2. Remain `Prefer` approach until # of endpoints hits 6
+
+**Scores on small dataset with different starting thresholds for local algorithm**
+|Algorithm/Starting threshold| Score |
+|:---|:---|
+|balanced|72.55|
+|local-1x|80.07|
+|local-2x|80.09|
+|local-3x|80.1|
+|local-4x|80.08|
+|local-5x|80.01|
+
+_Overload Maintenance:_
+
+Keep track of endpoints overload during endpoints rebalance, assign more endpoints to EndpointSlice Groups that are exceeding the overload threshold with a higher priority from zones:
+1. With negative overloads first
+2. With overloads less than overload threshold after zones in condition 1 runs out
+
+_Lazy Rebalance:_
+1. To reduce churn of EndpointSlices and overhead for the API server.
+2. After rebalancing endpoints among EndpointSlices, any following endpoints update without any endpoints exceeding a threshold, make that update only. 
+3. Do a full rebalance until the update can't be made without exceeding the threshold.
+
+###### Distribution of EndpointSlices
+Each EndpointSlice Group is consumed by a single zone and will only receive traffic from that zone (local zone). The number of endpoints belonging to that SliceGroup is calculated by `total endpoints * nodes in zone / total nodes` (**expected endpoints**). The control plane tries to fill SliceGroup with local endpoints as full as possible.
+1. Iterate through existing zones, assign all endpoints to its corresponding ‘local SliceGroups’. 
+2. If the current zone has enough endpoints to achieve traffic load below the overload threshold, put it into an ‘endpoints_available pool’. Otherwise put it into an ‘endpoints_require pool’. The pool is implemented by a priority queue which will place the zone with most extra endpoints or the zone with most insufficient endpoints first.
```

Comment:
Replace with backticks wherever this happens:
```suggestion
2. If the current zone has enough endpoints to achieve traffic load below the overload threshold, put it into an `endpoints_available` pool. Otherwise put it into an `endpoints_require` pool. The pool is implemented by a priority queue which will place the zone with most extra endpoints or the zone with most insufficient endpoints first.
```

Reply:
changed